#### Dictionaries
- dictionary is collection of unordered elements
- list is ordered
- set is unordered
- tuple is ordered


In [51]:
student={'name':'sireesha','branch':'ece'}
print(student)

{'name': 'sireesha', 'branch': 'ece'}


In [52]:
type(student)

dict

In [ ]:
print(type(student))

In [ ]:
li=[1,2,3,4]

In [54]:
len(student)

2

- min max here used in loxographical order that alphabatical order min is 'brach'  max is 'name'

In [55]:
min(student)

'branch'

In [56]:
max(student)

'name'

In [57]:
student.keys()

dict_keys(['name', 'branch'])

membership operator
- in 
- not in

In [58]:
'Branch' in student

False

In [59]:
'place' not in student

True

In [60]:
dir({})

['__class__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'clear',
 'copy',
 'fromkeys',
 'get',
 'items',
 'keys',
 'pop',
 'popitem',
 'setdefault',
 'update',
 'values']

In [61]:
student.__sizeof__()  # returns size of student dictionary in bytes 

216

In [62]:
student.__len__()

2

In [63]:
student

{'name': 'sireesha', 'branch': 'ece'}

In [64]:
student2=student
student2

{'name': 'sireesha', 'branch': 'ece'}

In [65]:
student2['name']='meloni'

In [66]:
student

{'name': 'meloni', 'branch': 'ece'}

In [67]:
student3=student.copy()

In [68]:
student3['name']='mallika'
student3

{'name': 'mallika', 'branch': 'ece'}

In [69]:
student

{'name': 'meloni', 'branch': 'ece'}

- by assigning 'student' to student1 if we made any changes in 'student1' and then 'student' will also change
- by copying the 'student' to 'student2' if we  made any changes in 'student2' nothing will change in 'student'

In [70]:
dict.fromkeys(range(10))

{0: None,
 1: None,
 2: None,
 3: None,
 4: None,
 5: None,
 6: None,
 7: None,
 8: None,
 9: None}

In [71]:
dict.fromkeys(range(3),0)

{0: 0, 1: 0, 2: 0}

In [72]:
dict.fromkeys(range(5),'new value')

{0: 'new value',
 1: 'new value',
 2: 'new value',
 3: 'new value',
 4: 'new value'}

In [73]:
nedict.fromkeys(range(5),'sireesha')

NameError: name 'nedict' is not defined

In [74]:
{1:'sirie',2:'name'}

{1: 'sirie', 2: 'name'}

In [77]:
{[1,2,3]:'sire',[1,4,3]:'name'}   # --> sicne list is mutable

TypeError: unhashable type: 'list'

In [78]:
{(1,12,3):'sir',(3,4,3):'fdkd'}   # --> since tuple is immutable 

{(1, 12, 3): 'sir', (3, 4, 3): 'fdkd'}

In [79]:
student.get('brach')

In [80]:
student

{'name': 'meloni', 'branch': 'ece'}

In [81]:
print(student.get('brach'))

None


In [82]:
print(student.get('place'))

None


In [83]:
# if we want any msg in place of None we can get if
print(student.get('place'),'absent')

None absent


In [84]:
print(student.get('place','absent'))

absent


In [85]:
# student.get() will return only that operation will not add place into student
student

{'name': 'meloni', 'branch': 'ece'}

In [86]:
student['brach']='cse'

In [87]:
student

{'name': 'meloni', 'branch': 'ece', 'brach': 'cse'}

In [88]:
del student['brach']

In [89]:
student

{'name': 'meloni', 'branch': 'ece'}

### dict.pop(key)
- removes the item related to the provided key

In [90]:
student.pop('adhar')

KeyError: 'adhar'

In [91]:
student['adhar']='3393839'

In [92]:
student

{'name': 'meloni', 'branch': 'ece', 'adhar': '3393839'}

In [93]:
student.pop('adhar')

'3393839'

In [94]:
student

{'name': 'meloni', 'branch': 'ece'}

In [95]:
student.popitem()  # --> branch and ece got eleminated

('branch', 'ece')

In [96]:
student

{'name': 'meloni'}

In [97]:
student.popitem()

('name', 'meloni')

In [98]:
student

{}

In [99]:
student.get('name','siri')  # Just returned a value

'siri'

In [100]:
student

{}

In [101]:
student.setdefault('name','ravi')

'ravi'

In [102]:
student

{'name': 'ravi'}

In [103]:
student.setdefault('name','raju')  # --> it has not updated because if a key is not present in the dic then 
# only it can add it key and value


'ravi'

In [104]:
student


{'name': 'ravi'}

In [105]:
student.keys()

dict_keys(['name'])

In [106]:
student.values()

dict_values(['ravi'])

In [107]:
student.items()

dict_items([('name', 'ravi')])

In [110]:
student2

{'name': 'ravi'}

In [111]:
student3

{'name': 'mallika', 'branch': 'ece'}

In [112]:
student3['place']='vizag'

In [113]:
student3

{'name': 'mallika', 'branch': 'ece', 'place': 'vizag'}

In [114]:
student

{'name': 'ravi'}

In [116]:
student.update(student3)
student

{'name': 'mallika', 'branch': 'ece', 'place': 'vizag'}

In [117]:
for key in student:   # --> we can only get key
    print(key)

name
branch
place


In [118]:
# if i want value
for key in student:
    print(student[key])

mallika
ece
vizag


In [121]:
for key in student:
    print(key,student[key],sep='----> ')

name----> mallika
branch----> ece
place----> vizag


In [122]:
for key in student:
    print(key,student[key],end=' ')

name mallika branch ece place vizag 

In [123]:
for key in student.keys():
    print(key,student[key],sep=' --> ')

name --> mallika
branch --> ece
place --> vizag


In [124]:
for key,value in student.items():
    print(key,value,sep=' --> ')

name --> mallika
branch --> ece
place --> vizag


In [125]:
for i,j in zip([1,2,3],[1,2,3]):
    print(i,j)

1 1
2 2
3 3


In [127]:
for i,j in zip([1,3,5],[4,7,9]):
    print(i,j)

1 4
3 7
5 9


In [128]:
for i,j in zip(['name','branch','place'],['sai','mech','visakhapatnam']):
    print(i,j)

name sai
branch mech
place visakhapatnam


In [129]:
s='hello world'
dic={}
for i in s:
    dic[i]=s.count(i)      # if we use count function for larger input we will get time complexity error
print(dic)

{'h': 1, 'e': 1, 'l': 3, 'o': 2, ' ': 1, 'w': 1, 'r': 1, 'd': 1}


In [132]:
s='hello world'
d={}
for i in s:
    if i not in d:
        d[i]=1
    else:
        d[i]+=1
print(d)

{'h': 1, 'e': 1, 'l': 3, 'o': 2, ' ': 1, 'w': 1, 'r': 1, 'd': 1}


In [135]:
freq={}
for char in s:
    freq[char]=freq.get(char,0)+1
    # dict[key] = old_value_or_zero +1
print(freq)

{'h': 1, 'e': 1, 'l': 3, 'o': 2, ' ': 1, 'w': 1, 'r': 1, 'd': 1}


#### find the frequency of each character from a to z in 
- s='hello world"
- o/p format---> {'a':0,'b':0,.....,'h':1,....,'z'=0}

In [142]:
s='hello world'
d={}
for num in range(ord('a'),ord('z')+1):
    #print(chr(num),end=' ')
    
    d[chr(num)]=s.count(chr(num))
print(d)
    

{'a': 0, 'b': 0, 'c': 0, 'd': 1, 'e': 1, 'f': 0, 'g': 0, 'h': 1, 'i': 0, 'j': 0, 'k': 0, 'l': 3, 'm': 0, 'n': 0, 'o': 2, 'p': 0, 'q': 0, 'r': 1, 's': 0, 't': 0, 'u': 0, 'v': 0, 'w': 1, 'x': 0, 'y': 0, 'z': 0}


In [155]:
s='hello world'
alphabets=list(map(chr,range(97,123)))
#alphabets
freq=dict.fromkeys(alphabets,0)
#freq
for char in s:
    if char in freq:
        freq[char]=freq.get(char,0)+1
print(freq) 


{'a': 0, 'b': 0, 'c': 0, 'd': 1, 'e': 1, 'f': 0, 'g': 0, 'h': 1, 'i': 0, 'j': 0, 'k': 0, 'l': 3, 'm': 0, 'n': 0, 'o': 2, 'p': 0, 'q': 0, 'r': 1, 's': 0, 't': 0, 'u': 0, 'v': 0, 'w': 1, 'x': 0, 'y': 0, 'z': 0}


In [156]:
ord('i')

105

In [149]:
ord('z')

122

### who is lucky whose number is lucky
- sai = 115+97+105  = 371 = 3+7+1 = 11 = 1+1 = 2
- ram = 114+97+109 = 374 = 3+7+4 = 14 = 1+4 = 5
- otuput is sai because he has the smallest number which is close to zero


In [9]:
s=['sai','ram']
l={}

for char in s:
    li=[]
    for i in range(len(char)):
        li.append(ord(char[i]))
    l[char]=li
print(li)
print(sum(li))

[114, 97, 109]
320


In [15]:

def sunn(s):
    v=0
    while(s!=0):
        r=s%10
        v+=r
        s=s//10
    if v>9:
        return sunn(v)
    else:
        return v

    
li=371
print(sunn(li))



2


In [10]:
n=371
s=0
while(n!=0):
    r=n%10
    #print(r)
    s=s+r
    n=n//10
print(s)

11


In [1]:
# sum of digits
l=[]
def sum_digits(b):
    if(b==0):
        return 1
    dig=b%10
    l.append(dig)
    sum_digits(b//10)
n=int(input("Enter a number: "))
sum_digits(n)
print(sum(l))

Enter a number: 371
11


In [17]:
def asciiSum(name):
    return str(sum(list(map(ord,name))))

name='sai'
asciiSum(name)

'317'

In [78]:
def reduce2Digit(s):
    result= sum(list(map(int,s)))
    if result>9:
        return reduce2Digit(str(result))
    return result


reduce2Digit('371')


2

In [82]:
names=['sai','ram','raju']
luckyNames={}
for name in names:
    total=asciiSum(name)
    luckyNumber=reduce2Digit(total)
    luckyNames[name]= luckyNumber
    
luckyNames



{'sai': 2, 'ram': 5, 'raju': 2}

In [83]:
leastNumber = min(luckyNames.values())
leastNumber

2

In [84]:
for item in luckyNames.items():
    if item[1]==leastNumber:
        print(item[0],item[1])

sai 2
raju 2


#### files in python
- To store the digital information
- To work with files in python 3 steps are there

In [19]:
fh = open('Python/book.txt','r')

data = fh.read()
print(data)
fh.close()

Line 1
Line 2
Line 3
Line 4
Line 5


In [20]:
fh = open('Python/book2.txt','r')  # we got error because book2 file is not there in the filepath

data = fh.read()
print(data)
fh.close()

FileNotFoundError: [Errno 2] No such file or directory: 'Python/book2.txt'

#### write(string_data)

In [24]:
fh = open('Python/book2.txt','w')

data = 'hi hello namaste '
fh.write(data)
fh.close()

In [25]:
fh = open('Python/book2.txt','a')

data = 'tinnara '
fh.write(data)
fh.close()

In [26]:
fh = open('Python/book3.txt','a')

data = 'tinnara '
fh.write(data)
fh.close()

In [27]:
fh=open('Python/book.txt','r')
for each in fh:
    print(each)
fh.close()

Line 1

Line 2

Line 3

Line 4

Line 5


In [36]:
fh=open('Python/book.txt','r')
i=1
for each in fh:
    print(i,each)
    i+=1
fh.close()

1 Line 1

2 Line 2

3 Line 3

4 Line 4

5 Line 5


In [37]:
fh=open('Python/book.txt','r')
i=1
for each in fh:
    if i==3:
        print(i,each)
    i+=1
fh.close()

3 Line 3



In [46]:
# Write the content from book.txt to output.txt using python file concept
# reading data from book.txt
ch = open('Python/book.txt','r')
data=ch.read()
ch.close()

# writing data to output.txt
fh = open('Python/output.txt','w')
fh.write(data)
fh.close()

In [48]:
# in my output.txt there should only name index and names there sholdn't marks and marks column
fh_i= open('Python/book.txt','r')
fh_o= open('Python/output.txt','w')

for line in fh_i:
    word=line.split()[0]+'\n'
    fh_o.write(word)
fh_i.close()
fh_o.close()



In [69]:
# in output.txt file there should be each name and corresponding total marks of each student
fh=open('Python/book.txt','r')
mh=open('Python/output.txt','w')
i=1
for line in fh:
    data=fh.read().split('\n')
print(data)
# for row in data:
#     li=li.append(row[0])
# print(li)

    
fh.close()
mh.close()



['sai   80  70  60', 'ravi  70  60  70', 'sasi  60  60  60']


In [2]:
fh_i=open('Python/book.txt','r')
fh_o=open('Python/output.txt','w')
count=0
for line in fh_i:
    count +=1
    if count==1:
        continue
    name = line.split()[0]
    marks= sum(list(map(int,line.split()[1:])))  # ['100','200','300']
    fh_o.write(name+' '+str(marks)+'\n')
fh_i.close()
fh_o.close()


In [6]:
fh_i=open('Python/book.txt','r')
fh_o=open('Python/output_1.txt','w')
count=0
for Line in fh_i:
    count +=1
    if count==1:
        fh_o.write(Line.strip()+'\ttotal')
    else:
        marks= sum(list(map(int,Line.split()[1:])))  # ['100','200','300']
        fh_o.write('\n'+Line.strip()+'\t'+str(marks))
fh_i.close()
fh_o.close()

#### tell()
#### seek()

In [ ]:
fh=open('Python/')
print('i am at ',fh.tell())
data=fh.read()


#### fh.readline()

In [9]:
fh=open('Python/output.txt','r')

data = fh.readline()
data = fh.readline()
print(data)

fh.close()

ravi 200



In [10]:
fh=open('Python/output.txt','r')

data = fh.readline()
data = fh.readlines()
print(data)

fh.close()

['ravi 200\n', 'sasi 180\n']


In [11]:
fh = open('Python/output.txt','r')    #  we opened in read mode but we are doing operation in write mode so that's why 
# we got error
data = 'hi'
fh.write(data)
fh.close()

UnsupportedOperation: not writable

In [14]:
s=input()
print(s)

234222
234222


In [15]:
int(s)

234222

In [16]:
str(s)

'234222'